### PR TITLE
Network

### PR DESCRIPTION
**Pull request by bastiscode**
PR for new network branch

### CODE REVIEW

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+        cw, ch = cw // self.config["pool_size"], ch // self.config["pool_size"]
+        return cw, ch
+
+    def _add_layer(self, layers: List[nn.Module], in_filters: int, out_filters: int) -> None:
+        layers.append(nn.Conv2d(in_filters, out_filters,
+                                kernel_size=self.config["conv_kernel_size"],
+                                stride=self.config["conv_kernel_stride"],
+                                padding=self.config["conv_kernel_padding"]))
+        layers.append(nn.BatchNorm2d(out_filters))
+        layers.append(nn.ReLU())
```
Comment (LMZimmer) [inline autoPyTorch/pipeline/components/setup/network/backbone/image.py:35]: We probably want to allow different activations

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    UniformFloatHyperparameter,
+    UniformIntegerHyperparameter
+)
+from torch import nn
+from torch.nn import functional as F
+
+from autoPyTorch.pipeline.components.setup.network.backbone.base_backbone import BaseBackbone
+
+
+class ConvNetBackbone(BaseBackbone):
```
Comment (LMZimmer) [inline autoPyTorch/pipeline/components/setup/network/backbone/image.py:18]: There can also be a ConvNetBackbone for e.g. time series. So we might want to consider renaming this to something like `ConvNetImageBackbone` or `ConvNet2DBackbone` and creating an additional `ConvNetBackbone` that acts as a parent class for `ConvNetImageBackbone` and `ConvNetTimeSeriesBackbone`/`ConvNet1DBackbone` (of course only if there actually is something to share). The same goes for other network types that can be used for different modalities but only if it makes

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+              // self.config["conv_kernel_stride"]) + 1
+        cw, ch = cw // self.config["pool_size"], ch // self.config["pool_size"]
+        return cw, ch
+
+    def _add_layer(self, layers: List[nn.Module], in_filters: int, out_filters: int) -> None:
+        layers.append(nn.Conv2d(in_filters, out_filters,
+                                kernel_size=self.config["conv_kernel_size"],
+                                stride=self.config["conv_kernel_stride"],
+                                padding=self.config["conv_kernel_padding"]))
+        layers.append(nn.BatchNorm2d(out_filters))
```
Comment (LMZimmer) [inline autoPyTorch/pipeline/components/setup/network/backbone/image.py:34]: Could we add an attribute to the networks `self.bn_args={"eps"=1e-05, "momentum"=0.1}` and have these be added when initializing bn layers `nn.BatchNorm2d(out_filters, **self.bn_args)`?

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+                                                           lower=min_pool_size,
+                                                           upper=max_pool_size))
+        return cs
+
+
+class _DenseLayer(nn.Sequential):
+    def __init__(self, num_input_features: int, growth_rate: int, bn_size: int, drop_rate: float):
+        super(_DenseLayer, self).__init__()
+        self.add_module('norm1', nn.BatchNorm2d(num_input_features)),
+        self.add_module('relu1', nn.ReLU(inplace=True)),
```
Comment (LMZimmer) [inline autoPyTorch/pipeline/components/setup/network/backbone/image.py:105]: Same goes for BN args and activations here

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+        for i in range(1, max_num_blocks + 1):
+            layer_hp = UniformIntegerHyperparameter('layer_in_block_%d' % i,
+                                                    lower=min_num_layers,
+                                                    upper=max_num_layers)
+            cs.add_hyperparameter(layer_hp)
+
+            if i > min_num_blocks:
+                cs.add_condition(CS.GreaterThanCondition(layer_hp, blocks_hp, i - 1))
+
+        return cs
```
Comment (LMZimmer) [inline autoPyTorch/pipeline/components/setup/network/backbone/image.py:229]: We definitely want a wide resnet for image data with shake drop and shake shake regularizations as options (see https://github.com/automl/Auto-PyTorch/blob/master/autoPyTorch/components/networks/image/resnet.py). Does not have to be now, but if we don't add it now please create an issue so that we don't forget it

File: autoPyTorch/pipeline/components/setup/network/backbone/tabular.py [hunk 6]
```diff
@@ -0,0 +1,619 @@
...
+        )
+        cs.add_hyperparameters([max_units])
+
+        max_dropout = UniformFloatHyperparameter(
+            "max_dropout", lower=0.0, upper=1.0
+        )
+        cs.add_hyperparameters([max_dropout])
+        cs.add_condition(CS.EqualsCondition(max_dropout, use_dropout, True))
+
+        return cs
```
Comment (LMZimmer) [inline autoPyTorch/pipeline/components/setup/network/backbone/tabular.py:619]: We also want a DenseNet here at some point, perhaps even the one form AutoGluon (https://arxiv.org/pdf/2003.06505.pdf, Figure 1). Maybe add this to the issue about networks we want to still implement

File: autoPyTorch/pipeline/components/setup/network/backbone/time_series.py [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+# Code inspired by https://github.com/hfawaz/InceptionTime
+# Paper: https://arxiv.org/pdf/1909.04939.pdf
+class _InceptionBlock(nn.Module):
+    def __init__(self,
+                 n_inputs: int,
+                 n_filters: int,
+                 kernel_size: int,
+                 bottleneck: int = None):
+        super(_InceptionBlock, self).__init__()
+        self.n_filters = n_filters
```
Comment (LMZimmer) [inline autoPyTorch/pipeline/components/setup/network/backbone/time_series.py:26]: Some linebreaks would be nice ;)

**Review by LMZimmer — CHANGES_REQUESTED** [review]
This already looks pretty good, added some suggestions. If these are addressed, tests are running and some more documentation is there, this is ready to be merged

File: autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    def __init__(
+            self,
+            random_state: Optional[np.random.RandomState] = None,
+            **kwargs: Any
+    ):
+        super().__init__(
+            random_state=random_state,
+        )
+        self.config = kwargs
+        self._backbones = get_available_backbones()
```
[the remark this answers is not in the document]
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py:29]: will be fixed

File: autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        self._heads = get_available_heads()
+        self._backbones = get_available_backbones()
+        self._heads = get_available_heads()
+
+    @staticmethod
+    def get_properties(dataset_properties: Optional[Dict[str, str]] = None) -> Dict[str, Any]:
```
[the remark this answers is not in the document]
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py:35]: Its because its overriden from autoPyTorch component, there it is also Dict[str, Any].

File: autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+
+    @staticmethod
+    def get_hyperparameter_search_space(dataset_properties: Optional[Dict[str, str]] = None,
+                                        **kwargs: Any) -> ConfigurationSpace:
+        cs = ConfigurationSpace()
+        backbones: Dict[str, Type[BaseBackbone]] = get_available_backbones()
+        heads: Dict[str, Type[BaseHead]] = get_available_heads()
+
+        if dataset_properties is not None and "task_type" in dataset_properties:
+            # filter backbones and heads for those who support the current task type
```
[the remark this answers is not in the document]
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py:49]: will be fixed

File: autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+            a self.config that is network specific, and contains the additional
+            configuration hyperparameters to build a domain specific network
+        """
+        backbone_name = self.config["backbone"]
+        head_name = self.config["head"]
+        Backbone = self._backbones[backbone_name]
+        Head = self._heads[head_name]
+
+        backbone = Backbone(**{k.replace(backbone_name, "").replace(":", ""): v
```
[the remark this answers is not in the document]
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py:83]: youre right, more clean

File: autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+                               for k, v in self.config.items() if
+                               k.startswith(backbone_name)})
```
[the remark this answers is not in the document]
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py:85]: in which utils would you put this function?

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py (hunk 2 above)
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/backbone/image.py:18]: Will rename it and will add the `ConvNetTimeSeriesBackbone` to the issue about the networks we want to implement

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py (hunk 3 above)
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/backbone/image.py:34]: Do you mean that they should be in the search space or just as a fixed attribute?

File: autoPyTorch/pipeline/components/setup/network/backbone/time_series.py (hunk 7 above)
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/backbone/time_series.py:26]: fixed

File: autoPyTorch/pipeline/components/setup/network/backbone/tabular.py (hunk 6 above)
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/backbone/tabular.py:619]: see #32

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py (hunk 5 above)
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/backbone/image.py:229]: see #32

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py (hunk 2 above)
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/backbone/image.py:18]: se #32

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py (hunk 1 above)
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/backbone/image.py]: fixed

File: autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (LMZimmer) [reply autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py:83]: I just merged a PR that introduced /utils/common.py that would probably be the best place (https://github.com/LMZimmer/Auto-PyTorch_refactor/blob/development/autoPyTorch/utils/common.py)

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py (hunk 3 above)
Comment (LMZimmer) [reply autoPyTorch/pipeline/components/setup/network/backbone/image.py:40]: I would set it as a fixed attribute for now. Perhaps in the future we want to study that influence or allow the user to change it manually

File: autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/BackboneHeadNet.py]: done (see https://github.com/LMZimmer/Auto-PyTorch_refactor/blob/network/autoPyTorch/utils/common.py)

File: autoPyTorch/pipeline/components/setup/network/base_network.py [hunk 13]
```diff
@@ -21,6 +22,9 @@ def __init__(
         self.random_state = random_state
         self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu") if device is None else device
 
+        self._fit_requirements = [FitRequirement("input_shape", tuple),
+                                  FitRequirement("output_shape", tuple)]
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/setup/network/base_network.py:29]: In the new PR #34 this will fail as the type argument has been changed to supported_types which expects an iterable.

Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/base_network.py:29]: What type should it be then? I will change it

File: autoPyTorch/pipeline/components/setup/network/backbone/image.py (hunk 3 above)
Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/backbone/image.py]: added bn_args to the image networks

File: autoPyTorch/pipeline/components/setup/network/base_network.py (hunk 13 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/setup/network/base_network.py:29]: just an Iterable over the type. So for example here it can be ```[FitRequirement("input_shape", [tuple, ]),
                                  FitRequirement("output_shape", [tuple, ])]```

Comment (bastiscode) [reply autoPyTorch/pipeline/components/setup/network/base_network.py]: Fixed
